plugins/engine-configuration/api/src/main/java/org/pentaho/di/engine/configuration/api/MetaStoreRunConfigurationFactory.java

```diff
@@ -30,10 +32,22 @@
  */
 public abstract class MetaStoreRunConfigurationFactory implements RunConfigurationFactory {
 
+  Logger logger = LoggerFactory.getLogger( MetaStoreRunConfigurationFactory.class );
+
   protected CheckedMetaStoreSupplier metastoreSupplier;
 
   public MetaStoreRunConfigurationFactory( CheckedMetaStoreSupplier metastoreSupplier ) {
-    this.metastoreSupplier = metastoreSupplier;
+    if( metastoreSupplier == null){
```

Comment:
why are we handling a null metastoreSupplier here?

is this related to `RunConfigurationProviderFactoryManagerImpl.java`
```
 public List<RunConfigurationProvider> generateProviders() {
    return generateProviders( null );
  }
```
And then why we only selecting the first one returned from a list and then we assume there is at least one and access a potential null obhect.

Reply:
I had moved the logic from [DefaultRunConfigurationProvider()](https://github.com/pentaho/pentaho-kettle/pull/9966/files#diff-7f75f5396784eea66235a48945302208615d1cce3d6682ca14d13354dbd0919bL60) and [SparkRunConfigurationProvider()](https://github.com/pentaho/spark-execution/pull/2/commits/8120ba6bd7d01e6012743f21e17007a4b786acf8#diff-e4d7d4ec0bea4a5fc4c055e38ce3fb0f010a75f6ddc57f3b5a6b8f5bcd7d9492L55) up to RunConfigurationProviderFactoryManagerImpl, but then decided to move it to their parent class here. I can move it back to RunConfigurationProviderFactoryManagerImpl if that makes more sense.

Reply:
So, through multiple fixes based on the review comments, I could change this to:
`return generateProviders( MetaStoreConst::getDefaultMetastore );`

In fact... that's a one liner, I can just do it. I had, as stated, moved that logic here because it's the abstract parent of the Providers, and the logic had been in the child classes, but, it seems that wasn't really the best place for it to begin with, so, I'll make that change now...

I am making all of these changes in grouped commits, I can squash it all after reviews are in if we don't want to "Squash and merge" on GitHub (Duarte had mentioned some reason as to why this can be a problem, I don't recall)

Reply:
My stance, the RunConfigurationManger and related classes should never be managing the MetaStore it should be passed to us and calling the supplier #get().

`MetaStoreConst::getDefaultMetastore` is a better approach than passing` null`.

I'd have to look at the code and flow of logic. to have a better recommendation than `MetaStoreConst::getDefaultMetastore`

Reply:
@singletonc  is there ever a situation where you won't have access to bowl or metastore either (abstractMeta, jobMeta, TransMeta) ?

I haven't done an exhaustive looks, But is the useds of RunConfigurationManager#getInstance()

- RunConfigurationImportExtensionPoint.java,  
 - you could access the RunConfigurationManager#getInstance(bowl) in side #callExtensionPoint(logChannelInterface, o)
- RunConfigurationSaveExtensionPoint.java
  - same thing as above
- RunConfigurationDelegate.java, (I assume UI selection)
 - #edit  calls  RunConfigurationManager logic
 - but another function #updateLoadedJobs has `spoonSupplier.get().getLoadedJobs()` , I assume we could get the bowl/metastore from this call or if there is a function related to "current focus job/trans"

Reply:
In conversation with @rmansoor, he thinks that if:
- what we have currently *works* with the code we currently have
- there is _not_ something drastically wrong with this approach

Then we ought to proceed with the current changes (that I have updated to this point based on review feedback) and we can refactor further as we approach the point at which we will be merging spark-execution into master. That then also answers the question we discussed today @NJtwentyone: our team will be doing this refactoring as project profile will be merging before us, we will rebase on them.